[Excel] (copilot) Sample combination add-in and agent (#968)
OfficeDev/Office-Add-in-samples: Samples/excel-copilot-agent/README.md

---
page_type: sample
urlFragment: excel-copilot-agent
products:
  - office-excel
  - office
  - m365
  - Teams
  - TeamsToolkit
languages:
  - javascript
  - typescript
extensions:
  contentType: samples
  technologies:
    - Add-ins
  createdDate: 05/15/2025 8:00:00 PM
description: "Shows how to a custom Copilot agent to analyze an Excel spreadsheet."
---

# Create data analysis charts in Excel with a Copilot agent

## Summary

This sample shows how to include a custom Copilot agent in an Excel add-in, and use it to insert charts that plot one column of data against another. 

## Description

> The user begins my opening the task pane and using it to insert sample data. Then the user opens Copilot, and opens the **Chart Analyze Agent**. Using natural language, the user asks Copilot to insert into the open worksheet a chart that plots the data of one column in a table against the data of another column. The user specifies the columns, but doesn't have to use the precise column names in the table.

Reviewer: Some mild grammar and typo suggestions. Also, consider adding an overarching sentence to set the purpose of the add-in.
```suggestion
This add-in lets the user create charts using Copilot's natural language processing. The user begins by opening the task pane and using it to insert sample data. Then, the user opens Copilot and opens the **Chart Analyze Agent**. Using natural language, the user asks Copilot to insert into the open worksheet a chart that plots the data of one column in a table against the data of another column. The user specifies the columns but doesn't have to use the precise column names in the table.
```

![Chart Analyze Agent pane with inserted chart](./resources/chart-analyse-agent-and-add-in.png)

## Applies to

- Excel on Windows
- Excel on the web

## Prerequisites

- Office connected to a Microsoft 365 subscription (including Office on the web).
- A Teams subscription, if it is not included in your Microsoft 365 subscription.
- [Node.js](https://nodejs.org/) version 16 or greater.
- [npm](https://docs.npmjs.com/downloading-and-installing-node-js-and-npm) version 8 or greater.
- Visual Studio Code

> = Teams Toolkit extension for Visual Studio Code

Reviewer: ```suggestion
- [Teams Toolkit extension for Visual Studio Code](https://marketplace.visualstudio.com/items?itemName=TeamsDevApp.ms-teams-vscode-extension)
```

## Solution

| Solution | Authors |
|----------|-----------|
| How to insert charts with natural language agents | Microsoft |

## Version history

| Version  | Date | Comments |
|----------|------|----------|
| 1.0 | 05-16-2025 | Initial release |

## Install the sample

1. Clone or download this repo.

1. Go to the **Samples\excel-copilot-management** folder and open it in Visual Studio Code.

1. In command prompt, bash shell, or **TERMINAL** in Visual Studio Code, navigate to the **Samples\excel-copilot-management** folder and run `npm install`.

1. Open **Teams Toolkit** from the **Activity Bar** in Visual Studio Code.

1. In the **ACCOUNTS** pane, ensure that support is enabled for both Copilot and uploading custom apps.

   ![Account pane in the toolkit](./resources/account-pane.png)

## Sideload the sample and start the server

1. In the **LIFECYCLE** pane, select **Provision**. This sideloads the add-in to Microsoft 365, including Excel.

1. In a command prompt, bash shell, or **TERMINAL** in Visual Studio Code at the root of the project, run `npm run dev-server` to start the server on localhost. 

>    **Note**: If you are prompted to delete an old certificate and/or to install a new one, agree to both prompts.

Reviewer: ```suggestion
   **Note**: If you are prompted to delete an old certificate or to install a new one, agree to both prompts.
```

   Wait until a line that is similar to the following appears at the bottom of the server window before continuing.

   ```console
   webpack 5.99.8 compiled successfully in 102 ms
   ```

## Try it out

1. The first step in testing depends on the platform.

>     - To test in Office on Windows, open Excel, and then open (or create) a workbook.

Reviewer: Consider highlighting the host to better guide the reader's eyes.
```suggestion
    - **Office on Windows**: Open Excel, and then open (or create) a workbook.
```

>     - To test in Office on the web, in a browser, navigate to `https://excel.cloud.microsoft`, and then open (or create) a workbook.

Reviewer: ```suggestion
    - **Office on the web**: In a browser, navigate to `https://excel.cloud.microsoft`, and then open or create a workbook.
```

>     - **Office on the web**: In a browser, navigate to `https://excel.cloud.microsoft`, and then open or create a workbook.

Reviewer: When sideloading on the web, the ribbon icon didn't appear:

![image](https://github.com/user-attachments/assets/e8c3e262-50a0-480a-91cb-8b6f983c14af)

> 1. Wait until the add-in has loaded. This may take as much as two minutes. The **Show Task Pane** button should appear on the **Home** ribbon automatically. If it doesn't, you need to manually activate the add-in: Select the **Add-ins** button on the **Home** ribbon, and then in the flyout that opens, select the **Chart Analyze** add-in. 

Reviewer: For some reason "much" doesn't sound to me like a measurement of time. 
```suggestion
1. Wait until the add-in has loaded. This may take as long as two minutes. The **Show Task Pane** button should appear on the **Home** ribbon automatically. If it doesn't, you need to manually activate the add-in: Select the **Add-ins** button on the **Home** ribbon, and then in the flyout that opens, select the **Chart Analyze** add-in. 
```

> 1. Wait until the add-in has loaded. This may take as much as two minutes. The **Show Task Pane** button should appear on the **Home** ribbon automatically. If it doesn't, you need to manually activate the add-in: Select the **Add-ins** button on the **Home** ribbon, and then in the flyout that opens, select the **Chart Analyze** add-in. 

Reviewer: My add-in never loaded in the web. Is there a step missing here? Feels like I need to give a sideload instruction to TTK.

Author: > My add-in never loaded in the web. Is there a step missing here? Feels like I need to give a sideload instruction to TTK.

All you need to do is the **Provision** in the toolkit. That sideloads to both Copilot and Excel. In Excel, try clicking the **Add-ins** button and seeing if the add-in is in the flyout.

1. On the **Home** tab and choose the **Show Task Pane** button in the ribbon to open it.

   ![Task pane of Chart Analyze add-in](./resources/chart-analyze-task-pane.png)

1. Select the **Initialize data** button on the task pane to insert a table of sample data into the workbook. Because this table gets focus, the **Table Design** ribbon opens. 

1. Switch back to the **Home** ribbon.

> 1. Open **Copilot** from the ribbon and select the hamburger control in the Copilot pane. The **Chart Analyze Agent** agent should be be in the list of agents. You may need to select **See more** to ensure that all agents are listed. If the agent isn't listed, try one or both of the following actions.

Reviewer: A minor thing, but should the agent be named "Chart Analysis"? Or is it meant to be a verb form?

   - Wait a few minutes and reload Copilot.
   - With Copilot open to the list of agents, click the cursor on the Copilot window and press Ctrl+R.

   ![List of agents in Copilot with Chart Analyze Agent circled in red.](./resources/copilot-agent-list.png)

1. When the agent is listed, select it and the pane for the agent opens.

   ![Chart Analyze Agent pane](./resources/chart-analyze-agent-pane.png)

> 1. Select the **Show me a chart that correlates sales and temperature** conversation starter, and then press the arrow head (send) control in the conversation box at the bottom of the pane. After a few moments, you are prompted to confirm. Select **Confirm**. A chart is inserted onto the worksheet. See the screen shot at the top of this readme file.

Reviewer: ```suggestion
1. Select the **Show me a chart that correlates sales and temperature** conversation starter, and then press the arrowhead (send) control in the conversation box at the bottom of the pane. After a few moments, you are prompted to confirm. Select **Confirm**. A chart is inserted onto the worksheet. See the screenshot at the top of this readme file.
```

1. Try entering natural language prompts for other combinations of columns in the table. 

## Make changes in the app

> Live reloading is supported for task pane portion of the app, but not the Copilot agent portion. To make changes, first shut down the server and uninstall the combind agent and add-in with these steps.

Reviewer: ```suggestion
Live reloading is supported for task pane portion of the app, but not the Copilot agent portion. To make changes, first shut down the server and uninstall the combined agent and add-in with these steps.
```

1. Close the Office application.

1. In the web server window press Ctrl+C. Choose "Y" in response to the prompt to end the process. 

   **Note**: If the web server doesn't shut down, in a command prompt, bash shell, or Visual Studio Code **TERMINAL** in the root of the project, run `npm run stop`.

> 1. Clear the Office cache following the instructions at [Manually clear the cache](https://learn.microsoft.com/en-us/office/dev/add-ins/testing/clear-cache#manually-clear-the-cache-in-excel-word-and-powerpoint).

Reviewer: ```suggestion
1. Clear the Office cache following the instructions at [Manually clear the cache](https://learn.microsoft.com/office/dev/add-ins/testing/clear-cache#manually-clear-the-cache-in-excel-word-and-powerpoint).
```

1. Open Teams and select **Apps** from the **Activity Bar**, then select **Manage your apps** at the bottom of the **Apps** pane.

1. Find the **Chart Analyze Agent** agent in the list of apps.

> 1. Select the arrow head to the left of the name to expand its row.

Reviewer: ```suggestion
1. Select the arrowhead to the left of the name to expand its row.
```
